scico/examples.py

```diff
@@ -0,0 +1,231 @@
+        f = open(os.path.join(temp_dir.name, zip_file), "wb")
+        f.write(data.read())
+        f.close()
+        if verbose:
+            print("Download complete")
+
+    # unzip downloaded data into temporary directory
+    for zip_file in (data_zip_files[channel], psf_zip_files[channel]):
+        if verbose:
+            print(f"Extracting content from zip file {zip_file}")
+        with zipfile.ZipFile(os.path.join(temp_dir.name, zip_file), "r") as zip_ref:
+            zip_ref.extractall(temp_dir.name)
+
+    # read unzipped data files into 3D arrays and save as .npz
+    zip_file = data_zip_files[channel]
+    y = volume_read(os.path.join(temp_dir.name, zip_file[:-4]))
+    zip_file = psf_zip_files[channel]
+    psf = volume_read(os.path.join(temp_dir.name, zip_file[:-4]))
+
+    npz_file = os.path.join(path, f"epfl_big_deconv_{channel}.npz")
+    if verbose:
+        print(f"Saving as {npz_file}")
+    np.savez(npz_file, y=y, psf=psf)
+
+
+def epfl_deconv_data(channel: int, verbose: bool = False, cache_path: str = None) -> JaxArray:
+    """Get deconvolution problem data from EPFL Biomedical Imaging Group.
+
+    If the data has previously been downloaded, it will be retrieved from
+    a local cache.
+
+    Args:
+        channel: Channel number between 0 and 2.
+        verbose: Flag indicating whether to print status messages.
+        cache_path: Directory in which downloaded data is cached. The
+           default is `~/.cache/scico/examples`, where `~` represents
+           the user home directory.
+
+    Returns:
+       tuple: A tuple (y, psf) containing:
+
+           - **y** : (DeviceArray): Blurred channel data.
+           - **psf** : (DeviceArray): Channel psf.
+    """
+
+    # set default cache path if not specified
+    if cache_path is None:
+        cache_path = os.path.join(os.path.expanduser("~"), ".cache", "scico", "examples")
+
+    # create cache directory and download data if not already present
+    npz_file = os.path.join(cache_path, f"epfl_big_deconv_{channel}.npz")
+    if not os.path.isfile(npz_file):
+        if not os.path.isdir(cache_path):
+            os.makedirs(cache_path)
+        get_epfl_deconv_data(channel, path=cache_path, verbose=verbose)
+
+    # load data and return y and psf arrays converted to float32
+    npz = np.load(npz_file)
+    y = npz["y"].astype(np.float32)
+    psf = npz["psf"].astype(np.float32)
+    return y, psf
+
+
+def downsample_volume(vol: JaxArray, rate: int) -> JaxArray:
+    """Downsample a 3D array.
+
+    Downsample a 3D array. If the volume dimensions can be divided by
+    `rate`, this is achieved via averaging distinct `rate` x `rate` x
+    `rate` block in `vol`. Otherwise it is achieved via a call to
+    :func:`scipy.ndimage.zoom`.
+
+    Args:
+        vol: Input volume.
+        rate: Downsampling rate.
+
+    Returns:
+        Downsampled volume.
+    """
+
+    if rate == 1:
+        return vol
+
+    if np.all([n % rate == 0 for n in vol.shape]):
```

Comment:
Interesting way of doing this. Is this the same as

https://scikit-image.org/docs/stable/api/skimage.transform.html#skimage.transform.downscale_local_mean

?

Reply:
It's the same when the volume shape is integer divisible by `rate`. Using `skimage.transform.downscale_local_mean` would be cleaner, but that would introduce an additional dependency on `skimage`.

Reply:
We agreed in the meeting not to add the `skimage` dependency.

Reply:
To be revisited if `skimage` becomes a dependency for other reasons.